### PR TITLE
Refactor `ConversionService` to remove mutable operations from `ConversionService.SHARED`

### LINKED ISSUE #7657: Do not allow registering shared state type convertor into the shared instance `ConversionService.DEFAULT`
### Issue description

There are some issues that are causing it https://github.com/micronaut-projects/micronaut-core/pull/7635

Major changes are needed:
- Remove methods that add converters to `ConversionService` and introduce `MutableConversionService`
- Allow contributing converters to `ConversionService.DEFAULT` only by service loader entries
- Register bean context converters to a managed instance of `ConversionService` that is delegating to `ConversionService.DEFAULT` just like Micronaut Data is doing https://github.com/micronaut-projects/micronaut-data/blob/a9c1f96ec45f3b4bf8dd250e8342e59e3b99d5b4/data-runtime/src/main/java/io/micronaut/data/runtime/convert/DataConversionServiceImpl.java

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes #7657

Hopefully replaced all the possible usages of `ConversionService.SHARED`.

Some use cases cannot work after this change:
The detached request body is being converted:
```
        when:
        def request = HttpRequest.POST('/', JsonObject.createObjectNode([:]))

        then:
        request.getBody(String).get() != '{}'
```

Not sure if it's something that is even make sense, the fix now it to use this:
```
        when:
        request.setConversionService(conversionService)

        then:
        request.getBody(String).get() == '{}'
```

### CODE REVIEW

**Comment by yawkat** [comment]
fyi this will conflict with #8100, I made a similar ConversionService change in DefaultHttpClient there

**Comment by graemerocher** [comment]
@dstepanov can you resolve the conflicts

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.function.Function;
+
+/**
+ * A version of {@link ConversionService} that supports adding new converters.
+ *
+ * @author Denis Stepanov
+ * @since 4.0.0
+ */
+public interface MutableConversionService extends ConversionService {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:28]: add nullability annotations to methods

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java (hunk 1 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:28]: Done

**Comment by dstepanov** [comment]
I think we can remove concurrent maps in the default implementation. I have measured it can get us 50% more throughput. The idea would be only to allow adding a new converter via `TypeConverterRegistrar`.

File: aop/src/main/java/io/micronaut/aop/InterceptedMethod.java [hunk 2]
```diff
@@ -40,7 +41,18 @@ public interface InterceptedMethod {
...
+    }
+
+    /**
+     * Creates a new instance of intercept method supporting intercepting different reactive invocations.
+     *
+     * @param context           The {@link MethodInvocationContext}
+     * @param conversionService The conversion service
+     * @return The {@link InterceptedMethod}
+     */
+    static InterceptedMethod of(MethodInvocationContext<?, ?> context, ConversionService conversionService) {
```
Comment (graemerocher) [inline aop/src/main/java/io/micronaut/aop/InterceptedMethod.java:54]: add `@since`

Comment (graemerocher) [reply aop/src/main/java/io/micronaut/aop/InterceptedMethod.java:54]: add nullability annotations

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java [hunk 3]
```diff
@@ -15,49 +15,24 @@
...
-     *
-     * @param sourceType    The source type
-     * @param targetType    The target type
-     * @param typeConverter The type converter
-     * @param <S>           The source generic type
-     * @param <T>           The target generic type
-     * @return This conversion service
-     */
-    <S, T> Impl addConverter(Class<S> sourceType, Class<T> targetType, TypeConverter<S, T> typeConverter);
+    ConversionService SHARED = new DefaultMutableConversionService();
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/convert/ConversionService.java:35]: Should we deprecate this field?

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Adds a type converter.
+     *
+     * @param sourceType    The source type
+     * @param targetType    The target type
+     * @param typeConverter The type converter
+     * @param <S>           The source generic type
+     * @param <T>           The target generic type
+     */
+    <S, T> void addConverter(@NonNull Class<S> sourceType, @NonNull Class<T> targetType, @NonNull Function<S, T> typeConverter);
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:50]: Return `MutableConversionService` to make the API fluent

File: http/src/main/java/io/micronaut/http/converters/SharedHttpConvertersRegistrar.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Optional;
+
+/**
+ * Converter registrar for HTTP classes.
+ *
+ * @author Denis Stepanov
+ * @since 4.0.0
+ */
+@Internal
+public class SharedHttpConvertersRegistrar implements TypeConverterRegistrar {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/converters/SharedHttpConvertersRegistrar.java:39]: make `final`. Does this need to be public?

File: inject/src/main/java/io/micronaut/context/ApplicationContext.java [hunk 6]
```diff
@@ -59,11 +58,6 @@
  */
 public interface ApplicationContext extends BeanContext, PropertyResolver, PropertyPlaceholderResolver {
 
-    /**
-     * @return The default conversion service
-     */
-    @NonNull ConversionService<?> getConversionService();
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/context/ApplicationContext.java:65]: Why was this deleted?

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java [hunk 7]
```diff
@@ -15,49 +15,24 @@
...
 
 /**
  * A service for allowing conversion from one type to another.
  *
- * @param <Impl> The type
  * @author Graeme Rocher
  * @since 1.0
  */
-public interface ConversionService<Impl extends ConversionService> {
+public interface ConversionService {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/convert/ConversionService.java:30]: I'm concerned about the removal of the generic type argument here as it likely to break a lot of end user code. Is this a necessary change? What is the justification?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java (hunk 3 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/ConversionService.java:35]: No, you can still access the convertors registered using the service loader.

File: http/src/main/java/io/micronaut/http/converters/SharedHttpConvertersRegistrar.java (hunk 5 above)
Comment (dstepanov) [reply http/src/main/java/io/micronaut/http/converters/SharedHttpConvertersRegistrar.java:39]: I think it needs to be public for the service loader.

File: inject/src/main/java/io/micronaut/context/ApplicationContext.java (hunk 6 above)
Comment (dstepanov) [reply inject/src/main/java/io/micronaut/context/ApplicationContext.java:65]: It's now defined in `ConversionServiceProvider`, which the context extends.

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java (hunk 7 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/ConversionService.java:30]: It was only used for the mutable `add` methods, now those methods are gone so the generic parameter is unused.

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java (hunk 4 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:50]: Do you want to have it generic like it was before?

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java (hunk 7 above)
Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/convert/ConversionService.java:30]: It does mean though that any place a user injected `ConversionService<?>` will break which might be a significant breaking change. Thoughts @sdelamo ?

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java (hunk 4 above)
Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:50]: Makes sense

Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:50]: I'm looking at it again, and I don't like to have it in this way. There are zero usages of chaining in the current codebase. The registration API defines this signature: `void register(ConversionService<?> conversionService)` which prevents to do the chaining. IMHO the generic parameter only makes the code ugly, and the chaining is not very important in this case. I would prefer not to have it. WDYT @micronaut-projects/core-developers ?

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java (hunk 7 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/convert/ConversionService.java:30]: Users might need to modify their code anyway if relying on the shared conversion service.

File: core/src/main/java/io/micronaut/core/convert/MutableConversionService.java (hunk 4 above)
Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/convert/MutableConversionService.java:50]: It is probably fine without it but the generic argument remains a potentially large source of breaking changes

File: core/src/main/java/io/micronaut/core/convert/ConversionService.java (hunk 7 above)
Comment (alvarosanchez) [reply core/src/main/java/io/micronaut/core/convert/ConversionService.java:30]: Our docs use `ConversionService<?>` in examples. Still I never liked that generic type argument, so my +1 to remove it if it's not needed

**Review by graemerocher — APPROVED** [review]
